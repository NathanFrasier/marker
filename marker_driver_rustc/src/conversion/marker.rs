//! This module and its sub modules form the translation layer from rustc's
//! internal representation to markers representation. All conversion methods
//! are implemented as methods of the [`MarkerConverterInner`] to group them
//! together and share access to common objects easily.

mod common;
mod expr;
mod generics;
mod item;
mod pat;
mod stmts;
mod ty;

use std::cell::RefCell;

use crate::context::storage::Storage;
use marker_api::{
    ast::{
        expr::ExprKind,
        item::{Body, ItemKind},
        BodyId, Crate, ExprId, ItemId, Span, SymbolId,
    },
    lint::Level,
};
use rustc_hash::FxHashMap;
use rustc_hir as hir;

/// An interface to convert rustc's IR to marker types.
///
/// This is a wrapper for [`MarkerConverterInner`] which is responsible for the
/// actual conversion. The conversion code from [`MarkerConverterInner`] has certain
/// expectations when it comes to the internal state. Using this wrapper ensures,
/// that these expectations are always fulfilled.
pub struct MarkerConverter<'ast, 'tcx> {
    inner: MarkerConverterInner<'ast, 'tcx>,
}

impl<'ast, 'tcx> MarkerConverter<'ast, 'tcx> {
    pub fn new(rustc_cx: rustc_middle::ty::TyCtxt<'tcx>, storage: &'ast Storage<'ast>) -> Self {
        Self {
            inner: MarkerConverterInner::new(rustc_cx, storage),
        }
    }

    forward_to_inner!(pub fn to_lint_level(&self, level: rustc_lint::Level) -> Level);
    forward_to_inner!(pub fn to_item(&self, rustc_item: &'tcx hir::Item<'tcx>) -> Option<ItemKind<'ast>>);
    forward_to_inner!(pub fn to_body(&self, body: &hir::Body<'tcx>) -> &'ast Body<'ast>);
    forward_to_inner!(pub fn to_span(&self, rustc_span: rustc_span::Span) -> Span<'ast>);
    forward_to_inner!(pub fn to_crate(
        &self,
        rustc_crate_id: hir::def_id::CrateNum,
        rustc_root_mod: &'tcx hir::Mod<'tcx>,
    ) -> &'ast Crate<'ast>);
}

macro_rules! forward_to_inner {
    (pub fn $fn_name:ident(&self $(, $arg_name:ident: $arg_ty:ty)* $(,)?) -> $ret_ty:ty) => {
        pub fn $fn_name(&self $(, $arg_name: $arg_ty)*) -> $ret_ty {
            self.inner.$fn_name($($arg_name, )*)
        }
    };
}
use forward_to_inner;

struct MarkerConverterInner<'ast, 'tcx> {
    rustc_cx: rustc_middle::ty::TyCtxt<'tcx>,
    storage: &'ast Storage<'ast>,

    // Converted nodes cache
    items: RefCell<FxHashMap<ItemId, ItemKind<'ast>>>,
    bodies: RefCell<FxHashMap<BodyId, &'ast Body<'ast>>>,
    exprs: RefCell<FxHashMap<ExprId, ExprKind<'ast>>>,
    num_symbols: RefCell<FxHashMap<u32, SymbolId>>,

    /// Lang-items are weird, and if I'm being honest, I'm uncertain that I
    /// completely understand them. Anyways, here it goes, this is my current
    /// understanding:
    ///
    /// User written paths like `String` are always stored as `Resolved` or
    /// `TypeRelative`. The `QPath::LangItem` is used for code that is
    /// generated by the compiler, for instance from `format_args!()`. This
    /// allows the compiler to add paths, without the need to resolve them,
    /// since the ID of lang-items is known. However, this makes it more
    /// complicated for marker, since there is no real path to convert. Also,
    /// there seems to be no real way to transform a lang-item to the path of
    /// the target item. This means manual mapping in marker... FUN...
    ///
    /// This is a map, connecting the lang-items with the [`SymbolId`]s which
    /// would correspond to the path. This mapping doesn't cover all items, as
    /// some should never be mapped but converted to other representations.
    /// (Adding all items at once would also not be too much fun)
    /// Also, lang items are common enough, to have this map filled by default,
    /// almost every `format_args` call uses them.
    ///
    /// The map is filled in [`Self::fill_create_lang_item_map`].
    lang_item_map: RefCell<FxHashMap<hir::LangItem, SymbolId>>,

    // Context information
    /// This holds the [`hir::BodyId`] of the body that is currently being
    /// converted. This may be [`None`] for items, but should always be [`Some`]
    /// for expressions, since they can (AFAIK) only occur inside bodies.
    /// Individual expressions can be requested via the driver context, however,
    /// this driver only provides IDs of converted expressions, meaning that
    /// the requested expression would be returned from cache and not
    /// require additional translations.
    rustc_body: RefCell<Option<hir::BodyId>>,
    /// Requested on demand from rustc using a [`hir::BodyId`] see
    /// [`MarkerConverterInner::rustc_body`] for more information
    rustc_ty_check: RefCell<Option<&'tcx rustc_middle::ty::TypeckResults<'tcx>>>,
}

// General util functions
impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
    fn new(rustc_cx: rustc_middle::ty::TyCtxt<'tcx>, storage: &'ast Storage<'ast>) -> Self {
        let s = Self {
            rustc_cx,
            storage,
            items: RefCell::default(),
            bodies: RefCell::default(),
            exprs: RefCell::default(),
            num_symbols: RefCell::default(),
            lang_item_map: RefCell::default(),
            rustc_body: RefCell::default(),
            rustc_ty_check: RefCell::default(),
        };

        s.fill_create_lang_item_map();

        s
    }

    /// See [`Self::lang_item_map`] for more context.
    fn fill_create_lang_item_map(&self) {
        use rustc_span::symbol::Symbol;
        #[rustfmt::skip]
        let list = [
            (hir::LangItem::FormatArguments, self.to_symbol_id(Symbol::intern("FormatArguments"))),
            (hir::LangItem::FormatArgument, self.to_symbol_id(Symbol::intern("FormatArgument"))),
        ];

        self.lang_item_map.borrow_mut().extend(list);
    }

    #[must_use]
    fn alloc<T>(&self, t: T) -> &'ast T {
        self.storage.alloc(t)
    }

    #[must_use]
    fn alloc_slice<T, I>(&self, iter: I) -> &'ast [T]
    where
        I: IntoIterator<Item = T>,
        I::IntoIter: ExactSizeIterator,
    {
        self.storage.alloc_slice(iter)
    }
}

impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
    #[must_use]
    fn to_crate(
        &self,
        rustc_crate_id: hir::def_id::CrateNum,
        rustc_root_mod: &'tcx hir::Mod<'tcx>,
    ) -> &'ast Crate<'ast> {
        self.alloc(Crate::new(
            self.to_crate_id(rustc_crate_id),
            self.to_items(rustc_root_mod.item_ids),
        ))
    }
}

pub fn create_lang_item_map() {}
