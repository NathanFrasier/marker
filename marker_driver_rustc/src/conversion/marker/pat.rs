use marker_api::ast::pat::{
    CommonPatData, IdentPat, OrPat, PatKind, RefPat, RestPat, SlicePat, StructFieldPat, StructPat, TuplePat,
    UnstablePat, WildcardPat,
};
use rustc_hir as hir;

use super::MarkerConversionContext;

impl<'ast, 'tcx> MarkerConversionContext<'ast, 'tcx> {
    #[must_use]
    pub fn to_pat(&self, pat: &hir::Pat<'tcx>) -> PatKind<'ast> {
        // Here we don't need to take special care for caching, as marker patterns
        // don't have IDs and can't be requested individually. Instead patterns are
        // stored as part of their parent expressions or items. Not needing to deal
        // with caching makes this implementation simpler.
        let data = CommonPatData::new(self.to_span_id(pat.span));

        match &pat.kind {
            hir::PatKind::Wild => PatKind::Wildcard(self.alloc(|| WildcardPat::new(data))),
            hir::PatKind::Binding(hir::BindingAnnotation(by_ref, mutab), id, ident, pat) => {
                PatKind::Ident(self.alloc(|| {
                    IdentPat::new(
                        data,
                        self.to_symbol_id(ident.name),
                        self.to_var_id(*id),
                        matches!(mutab, rustc_ast::Mutability::Mut),
                        matches!(by_ref, hir::ByRef::Yes),
                        pat.map(|rustc_pat| self.to_pat(rustc_pat)),
                    )
                }))
            },
            hir::PatKind::Struct(qpath, fields, has_rest) => {
                let api_fields = self.alloc_slice_iter(fields.iter().map(|field| {
                    StructFieldPat::new(
                        self.to_span_id(field.span),
                        self.to_symbol_id(field.ident.name),
                        self.to_pat(field.pat),
                    )
                }));
                PatKind::Struct(
                    self.alloc(|| StructPat::new(data, self.to_path_from_qpath(qpath), api_fields, *has_rest)),
                )
            },
            hir::PatKind::TupleStruct(qpath, pats, dotdot) => {
                let ddpos = dotdot.as_opt_usize();
                let offset_pos = ddpos.unwrap_or(usize::MAX);
                let api_fields = self.alloc_slice_iter(pats.iter().enumerate().map(|(mut index, pat)| {
                    if index >= offset_pos {
                        index += offset_pos;
                    }
                    StructFieldPat::new(
                        self.to_span_id(pat.span),
                        self.to_symbol_id_for_num(u32::try_from(index).expect("a index over 2^32 is unexpected")),
                        self.to_pat(pat),
                    )
                }));
                PatKind::Struct(
                    self.alloc(|| StructPat::new(data, self.to_path_from_qpath(qpath), api_fields, ddpos.is_some())),
                )
            },
            hir::PatKind::Or(pats) => PatKind::Or(
                self.alloc(|| OrPat::new(data, self.alloc_slice_iter(pats.iter().map(|rpat| self.to_pat(rpat))))),
            ),
            hir::PatKind::Tuple(pats, dotdot) => {
                let pats = if let Some(rest_pos) = dotdot.as_opt_usize() {
                    let (start, end) = pats.split_at(rest_pos);
                    self.chain_pats(start, self.new_rest_pat(), end)
                } else {
                    self.alloc_slice_iter(pats.iter().map(|pat| self.to_pat(pat)))
                };
                PatKind::Tuple(self.alloc(|| TuplePat::new(data, pats)))
            },
            hir::PatKind::Box(_) => PatKind::Unstable(self.alloc(|| UnstablePat::new(data))),
            hir::PatKind::Ref(pat, muta) => {
                PatKind::Ref(self.alloc(|| RefPat::new(data, self.to_pat(pat), matches!(muta, hir::Mutability::Mut))))
            },
            hir::PatKind::Slice(start, wild, end) => {
                let elements = if let Some(wild) = wild {
                    self.chain_pats(start, self.to_pat(wild), end)
                } else {
                    assert!(end.is_empty());
                    self.alloc_slice_iter(start.iter().map(|pat| self.to_pat(pat)))
                };
                PatKind::Slice(self.alloc(|| SlicePat::new(data, elements)))
            },
            // These haven't been implemented yet, as they require expressions.
            // The pattern creation is tracked in #50
            hir::PatKind::Path(_) => todo!("{pat:#?}"),
            hir::PatKind::Lit(_) => todo!("{pat:#?}"),
            hir::PatKind::Range(_, _, _) => todo!("{pat:#?}"),
        }
    }

    fn chain_pats(
        &self,
        start: &[hir::Pat<'tcx>],
        ast_wild: PatKind<'ast>,
        end: &[hir::Pat<'tcx>],
    ) -> &'ast [PatKind<'ast>] {
        let start = start.iter().map(|pat| self.to_pat(pat));
        let middle = std::iter::once(ast_wild);
        let end = end.iter().map(|pat| self.to_pat(pat));
        let api_pats: Vec<_> = start.chain(middle).chain(end).collect();
        self.alloc_slice_iter(api_pats.into_iter())
    }

    fn new_rest_pat(&self) -> PatKind<'ast> {
        // This is a dummy span, it's dirty, but at least works for the mean time :)
        let data = CommonPatData::new(self.to_span_id(rustc_span::DUMMY_SP));
        PatKind::Rest(self.alloc(|| RestPat::new(data)))
    }
}
