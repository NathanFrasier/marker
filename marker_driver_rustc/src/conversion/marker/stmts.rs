use marker_api::ast::stmt::{LetStmt, StmtKind};
use rustc_hir as hir;

use super::MarkerConverterInner;

impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
    pub fn to_stmt(&self, stmt: &hir::Stmt<'tcx>) -> Option<StmtKind<'ast>> {
        match &stmt.kind {
            hir::StmtKind::Local(local) => Some(StmtKind::Let(self.alloc(self.to_let_stmt(local)))),
            hir::StmtKind::Item(item) => self.to_item_from_id(*item).map(StmtKind::Item),
            hir::StmtKind::Expr(expr) | hir::StmtKind::Semi(expr) => Some(StmtKind::Expr(self.to_expr(expr))),
        }
    }

    fn to_let_stmt(&self, local: &hir::Local<'tcx>) -> LetStmt<'ast> {
        LetStmt::new(
            self.to_span_id(local.span),
            self.to_pat(local.pat),
            local.ty.map(|ty| self.to_ty(ty)),
            local.init.map(|init| self.to_expr(init)),
            local.els.map(|els| self.to_expr_from_block(els)),
        )
    }
}
