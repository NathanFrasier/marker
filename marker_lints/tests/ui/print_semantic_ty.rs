// normalize-stderr-windows: "tests/ui/" -> "$$DIR/"

use core::fmt::Debug;
use core::marker::PhantomData;

static HELP_ITEM_TYPE_SEQUENCE: Option<AllowSync<(&[i32], [i32; 8])>> = None;
static HELP_ITEM_TYPE_POINTER: Option<AllowSync<(&'static str, *const i32, *mut i32)>> = None;
static HELP_ITEM_TYPE_COMPLEX: Option<
    AllowSync<(
        AliasTy,
        String,
        Option<String>,
        Vec<UnionItem>,
        Box<dyn Debug>,
        Box<dyn Iterator<Item = i32> + 'static>,
    )>,
> = None;

type AliasTy = Box<u32>;

pub union UnionItem {
    _f: f32,
    _i: i32,
}

pub struct AllowSync<T> {
    _data: PhantomData<T>,
}
unsafe impl<T> Sync for AllowSync<T> {}

trait InterestingTrait<T> {
    type A: Default;
    fn use_alias(&self) {
        // FIXME: This expression is currently not found by the print test code.
        // Try to figure out why and then make sure that is is correctly represented.
        let _ty: Self::A = Self::A::default();
    }
}

fn param_type<T: Debug>(t: T) {
    let _ty_generic: T = t;
}

fn main() {
    let _ty: u32 = 10;
    let _ty_primitive: Option<(u8, u16, u32, u64, u128, usize)> = None;
    let _ty_primitive: Option<(i8, i16, i32, i64, i128, isize)> = None;
    let _ty_primitive: Option<(char, bool, f32, f64)> = None;
    let _ty_sequence: [u32; 1] = [10];
    let slice: &[u32] = &[10];
    let _ty_sequence: &[u32] = slice;
    let _ty_ptr: Option<(&'static str, *const i32, *mut i32)> = None;

    // Interestingly, rustc substitutes the type directly and the semantic type
    // doesn't show the type alias.
    let _ty_simple_alias: AliasTy = AliasTy::new(12);

    let _ty_adt: String = String::new();
    let _ty_dyn_simple: Option<Box<dyn Debug>> = None;
    let _ty_dyn_complex: Option<Box<dyn Iterator<Item = i32> + 'static>> = None;
}
