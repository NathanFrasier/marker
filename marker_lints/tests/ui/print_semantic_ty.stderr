warning: print type test
  --> $DIR/print_semantic_ty.rs:28:5
   |
28 |     let _ty_generic: T = t;
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Generic(
                   SemGenericTy {
                       _lifetime: PhantomData<&()>,
                       generic_id: GenericId(..),
                   },
               ),
           }
   = note: `#[warn(marker::test_lint)]` on by default

warning: print type test
  --> $DIR/print_semantic_ty.rs:36:5
   |
36 |     let _ty: u32 = 10;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Num(
                   u32,
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:37:5
   |
37 |     let _ty_primitive: Option<(u8, u16, u32, u64, u128, usize)> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Tuple(
                                           (
                                               SemTy {
                                                   kind: Num(
                                                       u8,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       u16,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       u32,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       u64,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       u128,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       usize,
                                                   ),
                                               },
                                           ),
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:38:5
   |
38 |     let _ty_primitive: Option<(i8, i16, i32, i64, i128, isize)> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Tuple(
                                           (
                                               SemTy {
                                                   kind: Num(
                                                       i8,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       i16,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       i32,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       i64,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       i128,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       isize,
                                                   ),
                                               },
                                           ),
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:39:5
   |
39 |     let _ty_primitive: Option<(char, bool, f32, f64)> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Tuple(
                                           (
                                               SemTy {
                                                   kind: Text(
                                                       char,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Bool(
                                                       bool,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       f32,
                                                   ),
                                               },
                                               SemTy {
                                                   kind: Num(
                                                       f64,
                                                   ),
                                               },
                                           ),
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:40:5
   |
40 |     let _ty_sequence: [u32; 1] = [10];
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Array(
                   [
                       SemTy {
                           kind: Num(
                               u32,
                           ),
                       },
                   ],
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:42:5
   |
42 |     let _ty_sequence: &[u32] = slice;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Ref(
                   SemRefTy {
                       mutability: Unmut,
                       inner_ty: SemTy {
                           kind: Slice(
                               [
                                   SemTy {
                                       kind: Num(
                                           u32,
                                       ),
                                   },
                               ],
                           ),
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:43:5
   |
43 |     let _ty_ptr: Option<(&'static str, *const i32, *mut i32)> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Tuple(
                                           (
                                               SemTy {
                                                   kind: Ref(
                                                       SemRefTy {
                                                           mutability: Unmut,
                                                           inner_ty: SemTy {
                                                               kind: Text(
                                                                   str,
                                                               ),
                                                           },
                                                       },
                                                   ),
                                               },
                                               SemTy {
                                                   kind: RawPtr(
                                                       SemRawPtrTy {
                                                           mutability: Unmut,
                                                           inner_ty: SemTy {
                                                               kind: Num(
                                                                   i32,
                                                               ),
                                                           },
                                                       },
                                                   ),
                                               },
                                               SemTy {
                                                   kind: RawPtr(
                                                       SemRawPtrTy {
                                                           mutability: Mut,
                                                           inner_ty: SemTy {
                                                               kind: Num(
                                                                   i32,
                                                               ),
                                                           },
                                                       },
                                                   ),
                                               },
                                           ),
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:44:5
   |
44 |     let _ty_fn_item: fn(u32) -> f32 = u32_to_f32;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: FnTy(
                   SemFnTy {
                       fn_id: ItemId(..),
                       generics: SemGenericArgs {
                           args: [],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:49:5
   |
49 |     let _ty_fn_ptr: fn(u32) -> f32 = fn_ptr;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: FnPtr(
                   SemFnPtrTy {
                       safety: Safe,
                       abi: Default,
                       params: [
                           SemTy {
                               kind: Num(
                                   u32,
                               ),
                           },
                       ],
                       return_ty: SemTy {
                           kind: Num(
                               f32,
                           ),
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:53:5
   |
53 |     let _ty_simple_alias: AliasTy = AliasTy::new(12);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Num(
                                           u32,
                                       ),
                                   },
                               ),
                               Ty(
                                   SemTy {
                                       kind: Adt(
                                           SemAdtTy {
                                               def_id: TyDefId(..),
                                               generics: SemGenericArgs {
                                                   args: [],
                                               },
                                           },
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:55:5
   |
55 |     let _ty_adt: String = String::new();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:56:5
   |
56 |     let _ty_dyn_simple: Option<Box<dyn Debug>> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Adt(
                                           SemAdtTy {
                                               def_id: TyDefId(..),
                                               generics: SemGenericArgs {
                                                   args: [
                                                       Ty(
                                                           SemTy {
                                                               kind: TraitObj(
                                                                   SemTraitObjTy {
                                                                       bound: [
                                                                           SemTraitBound {
                                                                               is_relaxed: false,
                                                                               trait_id: TyDefId(..),
                                                                               trait_generic_args: SemGenericArgs {
                                                                                   args: [],
                                                                               },
                                                                           },
                                                                       ],
                                                                   },
                                                               ),
                                                           },
                                                       ),
                                                       Ty(
                                                           SemTy {
                                                               kind: Adt(
                                                                   SemAdtTy {
                                                                       def_id: TyDefId(..),
                                                                       generics: SemGenericArgs {
                                                                           args: [],
                                                                       },
                                                                   },
                                                               ),
                                                           },
                                                       ),
                                                   ],
                                               },
                                           },
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: print type test
  --> $DIR/print_semantic_ty.rs:57:5
   |
57 |     let _ty_dyn_complex: Option<Box<dyn Iterator<Item = i32> + 'static>> = None;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: SemTy {
               kind: Adt(
                   SemAdtTy {
                       def_id: TyDefId(..),
                       generics: SemGenericArgs {
                           args: [
                               Ty(
                                   SemTy {
                                       kind: Adt(
                                           SemAdtTy {
                                               def_id: TyDefId(..),
                                               generics: SemGenericArgs {
                                                   args: [
                                                       Ty(
                                                           SemTy {
                                                               kind: TraitObj(
                                                                   SemTraitObjTy {
                                                                       bound: [
                                                                           SemTraitBound {
                                                                               is_relaxed: false,
                                                                               trait_id: TyDefId(..),
                                                                               trait_generic_args: SemGenericArgs {
                                                                                   args: [
                                                                                       TyBinding(
                                                                                           SemTyBindingArg {
                                                                                               binding_target: ItemId(..),
                                                                                               ty: SemTy {
                                                                                                   kind: Num(
                                                                                                       i32,
                                                                                                   ),
                                                                                               },
                                                                                           },
                                                                                       ),
                                                                                   ],
                                                                               },
                                                                           },
                                                                       ],
                                                                   },
                                                               ),
                                                           },
                                                       ),
                                                       Ty(
                                                           SemTy {
                                                               kind: Adt(
                                                                   SemAdtTy {
                                                                       def_id: TyDefId(..),
                                                                       generics: SemGenericArgs {
                                                                           args: [],
                                                                       },
                                                                   },
                                                               ),
                                                           },
                                                       ),
                                                   ],
                                               },
                                           },
                                       ),
                                   },
                               ),
                           ],
                       },
                   },
               ),
           }

warning: 14 warnings emitted

